depdive/src/github.rs

```diff
@@ -165,20 +175,52 @@ impl GitHubAnalyzer {
             .unwrap_or(false)
     }
 
-    fn get_github_repo_fullname(repo_url: &Url) -> Result<String> {
+    fn get_github_repo_fullname(&self, repo_url: &Url) -> Result<String> {
+        // Return the repository full name if a valid GitHub url
+
         if !Self::is_github_url(repo_url) {
-            return Err(anyhow!("Repository, {}, is not from GitHub", repo_url));
+            return Err(InvalidGitHubRepoError {
+                url: repo_url.clone(),
+            }
+            .into());
         }
 
-        let mut segments = repo_url.path_segments().unwrap();
-        let owner = segments
-            .next()
-            .ok_or_else(|| anyhow!("repository url missing owner"))?;
+        let mut segments = repo_url
+            .path_segments()
+            .ok_or_else(|| InvalidGitHubRepoError {
+                url: repo_url.clone(),
+            })?;
+        let owner = segments.next().ok_or_else(|| InvalidGitHubRepoError {
+            url: repo_url.clone(),
+        })?;
         let repo = segments
             .next()
             .map(|repo| repo.trim_end_matches(".git"))
-            .ok_or_else(|| anyhow!("repository url missing repo"))?;
-        return Ok(format!("{}/{}", owner, repo));
+            .ok_or_else(|| InvalidGitHubRepoError {
+                url: repo_url.clone(),
+            })?;
+
+        let repo_fullname = format!("{}/{}", owner, repo);
+        let valid_flag = self.is_valid_github_repo(&repo_fullname)?;
```

Comment:
you may have to do a `map_err` here if to convert the error type to InvalidGitHubRepoError though

Reply:
Tried to differentiate between `repo not found` and other failure like `connection failed` here. Therefore, we return `anyhow` error and downcast above and treat the two errors differently.

Reply:
Then you should look at adding variants to your error type to differentiate the two. There should be examples of this in the thiserror crate documentation